src/main/java/com/metamx/common/io/smoosh/FileSmoosher.java

```diff
@@ -206,13 +231,132 @@ public void close() throws IOException
               String.format("Expected [%,d] bytes, only saw [%,d], potential corruption?", size, bytesWritten)
           );
         }
+        // Merge temporary files on to the main smoosh file.
+        mergeWithSmoosher();
+      }
+    };
+  }
+
+  /**
+   * Merges temporary files created by delegated SmooshedWriters on to the main
+   * smoosh file.
+   *
+   * @throws IOException
+   */
+  private void mergeWithSmoosher() throws IOException
+  {
+    // Get processed elements from the stack and write.
+    List<File> fileToProcess = new ArrayList<>(completedFiles);
```

Comment:
Maybe it can be just iteration over completedFiles (currently lines 251-256), then `completedFiles.clear()`.